### DISCUSSION TITLE
saving segmentation with original spacing

### DISCUSSION THREAD

**Discussion by Meddebma** [Project-MONAI/MONAI#2668] [General]
Hi, 
I much appreciated the newly added original image spacing to the spleen tutorial:

https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d.ipynb

I had the problem until now that the saved the segmentation masks did not fit to original images in the Nifti Reader. 
Can you please tell me (or add it in the tutorial) how to save the segmentations with the original image spacing?

thanks

**Comment by Nic-Ma** [comment]
Hi @Meddebma ,

Thanks for your interest here.
Please refer to this example tutorial for saving after inverting:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_inference_dict.py#L88

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma, 

thanks for your answer it worked perfectly, I get though two channels (Segmentation and background), is it possible to save only the segmentation result? 
I saw there is a Parameter squeeze_end_dims in SaveImaged

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

You can add a `SplitChanneld` transform to split the data.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma, 
so I tried this:
post_transforms = Compose([
    EnsureTyped(keys="pred"),
    Invertd(
        keys="pred",
        transform=val_org_transforms,
        orig_keys="image",
        meta_keys="pred_meta_dict",
        orig_meta_keys="image_meta_dict",
        meta_key_postfix="meta_dict",
        nearest_interp=False,
        to_tensor=True,
    ),
    KeepLargestConnectedComponentd(keys="pred",applied_labels=[1]),
    AsDiscreted(keys="pred", argmax=True, to_onehot=True, n_classes=2),
    AsDiscreted(keys="label", to_onehot=True, n_classes=2),
    **SplitChanneld(keys="pred", channel_dim=0),**
    SaveImaged(keys="pred", meta_keys="pred_meta_dict", output_dir="/content/drive/My Drive/Spleen_AI/CH Data/mask", output_postfix="seg", resample=False)
])
It did not change anything, then I changed the channel_dim to 1, and I've got also two channels saved out

**Comment by Nic-Ma** [comment]
Hi @Meddebma ,

When splitting channels, the transform will save the results into different keys, so please set the key in `SaveImaged` accordingly instead of the original `pred`.
Refer to:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/utility/dictionary.py#L345

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma, 
thank you for your notice, 
I tried with Pred_0, I only get the Background, and when I put Pred_1 I get a completely black screen containing nothing :/

**Reply by Nic-Ma** [reply]
May I know what you mean "completely black screen"? Because it's 3D image, maybe you need to use some 3D viewer?
And what's the evaluation metrics with your model? Have you trained it to predict reasonable output?
And if you can share your full test program, @yiheng-wang-nv may help double check it.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma ,

actually the notebook works well and when I save the batch as it is I get a perfect segmentation mask, the problem does only occur when I use the splitchannel transform.
Here is my notebook:
https://github.com/Meddebma/pyradiomics/blob/master/Untitled1.ipynb

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

May I know how did you view the saved image? Please note that by default it will squeeze the channel dim if only 1 channel:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/io/dictionary.py#L185.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma,

I displayed the saved image in ITK-Snap and in the 3D Slicer. I attached here the original mask as I want to get it, and the saved output with two channels:
[Spleen_121_output.nii.gz](https://github.com/Project-MONAI/MONAI/files/6907211/Spleen_121_output.nii.gz)
[Spleen_121.nii.gz](https://github.com/Project-MONAI/MONAI/files/6907215/Spleen_121.nii.gz)

**Reply by Nic-Ma** [reply]
Hi @rijobro ,

I am not very familiar with the 3D view tools, do you have any idea why the split prediction is all black in viewer?
@Meddebma , is there a way to print out the saved raw data to see what's in the array?

Thanks.

**Reply by Meddebma** [reply]
Hi, 
sure here are the saved data diplayed in the notebook
https://github.com/Meddebma/pyradiomics/blob/master/Untitled1.ipynb

**Reply by Nic-Ma** [reply]
Could you please help try again with below postprocessing transforms?
```py

post_transforms = Compose([
    EnsureTyped(keys="pred"),
    Invertd(
        keys="pred",
        transform=val_org_transforms,
        orig_keys="image",
        meta_keys="pred_meta_dict",
        orig_meta_keys="image_meta_dict",
        meta_key_postfix="meta_dict",
        nearest_interp=False,
        to_tensor=True,
    ),
    AsDiscreted(keys="pred", argmax=True),
    SaveImaged(keys="pred", meta_keys="pred_meta_dict", output_dir="/content/drive/My Drive/Spleen_AI/CH Data/mask1", output_postfix="seg", resample=False)
])
```
I checked your code and I think you don't need to change the prediction to One-Hot after Argmax because you don't compute metrics.

Thanks.

**Reply by Meddebma** [reply]
Dear @Nic-Ma 
thank you so much it worked perfectly and the saved output correspond exactly to the original image in the Nifti Readers. Appreciate your continuous help and tremendous work with MONAI!

**Reply by Nic-Ma** [reply]
NP, glad to see your good progress!
Thanks.
